src/Builder.jl

```diff
@@ -76,32 +76,99 @@ Selectors.strict{T <: DocumentPipeline}(::Type{T}) = false
-                src = normpath(joinpath(root, file))
-                dst = normpath(joinpath(build, relpath(root, source), file))
-                # Non-markdown files are simply copied over to `build`.
-                # Markdown files get added to the document tree as `Page` objects.
-                endswith(src, ".md") ?
-                    Documents.addpage!(doc, src, dst) :
-                    cp(src, dst; remove_destination = true)
+
+    # We'll walk over all the files in the .user.source directory.
+    # The directory structure is copied over to .user.build. All files, with
+    # the exception of markdown files (identified by the extension) are copied
+    # over as well, since they're assumed to be images, data files etc.
+    # Markdown files, however, get added to the document and also stored into
+    # `mdpages`, to be used later.
+    mdpages = Compat.String[]
+    for (root, dirs, files) in walkdir(source)
+        for dir in dirs
+            d = normpath(joinpath(build, relpath(root, source), dir))
+            isdir(d) || mkdir(d)
+        end
+        for file in files
+            src = normpath(joinpath(root, file))
+            dst = normpath(joinpath(build, relpath(root, source), file))
+            if endswith(file, ".md")
+                push!(mdpages, Utilities.srcpath(source, root, file))
+                Documents.addpage!(doc, src, dst)
+            else
+                cp(src, dst; remove_destination = true)
             end
         end
-    else
-        error("source directory '$(abspath(source))' is missing.")
     end
+
+    # If the user hasn't specified the page list, then we'll just default to a
+    # flat list of all the markdown files we found, sorted by the filesystem
+    # path (it will group them by subdirectory, among others).
+    userpages = isempty(doc.user.pages) ? sort(mdpages) : doc.user.pages
+
+    # Populating the .navtree and .navlist.
+    # We need the for loop because we can't assign to the fields of the immutable
+    # doc.internal.
+    for navnode in walk_navpages(userpages, nothing, doc)
+        push!(doc.internal.navtree, navnode)
+    end
+
+    # Finally we populate the .next and .prev fields of the navnodes that point
+    # to actual pages.
+    local prev::Nullable{Documents.NavNode} = nothing
+    for navnode in doc.internal.navlist
+        navnode.prev = prev
+        Utilities.unwrap(prev) do prevnode
+            prevnode.next = navnode
+        end
+        prev = navnode
+    end
+end
+
+"""
+    walk_navpages(x, parent, doc)
+
+Recursively walks through the [`Documents.Document`](@ref)'s `.user.pages` field,
+generating [`Documents.NavNode`](@ref)s and related data structures in the
+process.
+
+This implementation is the de facto specification for the `.user.pages` field.
+"""
+walk_navpages(ps::Vector, parent, doc) = [walk_navpages(p, parent, doc)::Documents.NavNode for p in ps]
+# NOTE (Compat): the `T <: Compat.String` is required for 0.4 compatibility,
```

Comment:
Yes, that sounds better. It gets rid of all the type parameters, including the one that we would otherwise have even in 0.5.